url78eec0e0b6_Orikuro_vic_tgz-pJ8-net_nexon_vindictus_itemcomparer_inport_StaticImportJ8/src/net/nexon/vindictus/itemcomparer/logic/ThreadStarter.java

```diff
@@ -0,0 +1,311 @@
+		sb.insert(0, i_end2);
+		sb.insert(0, i_end1);
+		sb.insert(0, i_comp);
+		sb.insert(0, i_atk);
+		sb.insert(0, i_res);
+		sb.insert(0, i_anzahl);
+		sb.insert(0, PARAMETER + "\n");
+
+		if (combs.size() == 0) {
+			System.out.println("Nothing found, price was too low.");
+			return;
+		}
+
+		System.out.println(Combo.OUTPUT_HEAD);
+		for (int i = 0; i < combs.size(); i++) {
+			System.out.println(i + ";" + combs.get(i));
+			if (i >= 4) {
+				System.out.println("[...]");
+				break;
+			}
+		}
+
+		if (!nocsv) {
+			writeResults(combs, sb.toString(), noinfo);
+		}
+	}
+
+	private int[] calcBestSets(HashSet<ItemSet> sets) {
+
+		List<Integer> def = new ArrayList<>();
+		List<Integer> atk = new ArrayList<>();
+		List<Integer> matk = new ArrayList<>();
+
+		HashMap<Integer, String> i_s = new HashMap<>();
+		for (ItemSet set_a : sets) {
+			int def0 = set_a.getDefBonus(5);
+			def.add(def0);
+			i_s.put(def0, set_a + " 5");
+
+			int atk0 = (int) set_a.getAtkBonus(5);
+			atk.add(atk0);
+			i_s.put(atk0, set_a + " 5");
+
+			int matk0 = set_a.getMatkBonus(5);
+			matk.add(matk0);
+			i_s.put(matk0, set_a + " 5");
+
+			for (ItemSet set_b : sets) {
+				if (set_a.equals(set_b))
+					continue;
+				else {
+					int def1 = set_a.getDefBonus(2) + set_b.getDefBonus(2);
+					int def2 = set_a.getDefBonus(2) + set_b.getDefBonus(3);
+					def.add(def1);
+					i_s.put(def1, set_a + " 2 2 " + set_b);
+					def.add(def2);
+					i_s.put(def2, set_a + " 2 3 " + set_b);
+
+					int atk1 = (int) (set_a.getAtkBonus(2) + set_b
+							.getAtkBonus(2));
+					int atk2 = (int) (set_a.getAtkBonus(2) + set_b
+							.getAtkBonus(3));
+					atk.add(atk1);
+					i_s.put(atk1, set_a + " 2 2 " + set_b);
+					atk.add(atk2);
+					i_s.put(atk2, set_a + " 2 3 " + set_b);
+
+					int matk1 = set_a.getMatkBonus(2) + set_b.getMatkBonus(2);
+					int matk2 = set_a.getMatkBonus(2) + set_b.getMatkBonus(3);
+					matk.add(matk1);
+					i_s.put(matk1, set_a + " 2 2 " + set_b);
+					matk.add(matk2);
+					i_s.put(matk2, set_a + " 2 3 " + set_b);
+
+				}
+			}
+		}
+
+		int maxdef = Collections.max(def);
+		int maxatk = Collections.max(atk);
+		int maxmatk = Collections.max(matk);
+		int[] maximums = new int[3];
+		maximums[0] = maxdef;
+		maximums[1] = maxatk + 6; // to offset some rounding issues
+		maximums[2] = maxmatk;
+
+		System.out.println(Arrays.toString(maximums));
+		System.out.println("Def:  " + i_s.get(maxdef));
+		System.out.println("Atk:  " + i_s.get(maxatk));
+		System.out.println("Matk: " + i_s.get(maxmatk));
+
+		return maximums;
+
+	}
+
+	private void writeResults(List<Combo> combs, String infos, boolean noinfo) {
+		File out = new File("_out" + File.separator);
+		out.mkdirs();
+		if (combs == null || combs.size() < 1)
+			return;
+		try {
+			long time = System.currentTimeMillis();
+			BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(
+					new FileOutputStream("_out" + File.separator + time
+							+ ".csv"), "UTF-8"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the resource created by the following constructor: *OutputStreamWriter*. The resource is referenced at lines 285, 288, 289, 292, 293. The resource is closed at line 295. There are other execution paths that do not contain closure statements, for example, when BufferedWriter.newLine throws an exception. Ensure the object returned by *OutputStreamWriter()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).